design-documents/promises.md

```diff
@@ -0,0 +1,112 @@
+### Why?
+Service contracts in the future will often be executed in an asynchronous manner
+and it's time to introduce a standard Promise to Magento for asynchronous operations to employ
+### Requirements
+* Promises CANNOT be forwarded with _then_ and _otherwise_ ([see explanation](#forwarding))
+* _then_ accepts a function that will be executed when the promise is resolved, the callback will
+receive a single argument - result of the execution
+* _otherwise_ accepts a function that will be executed if an error occurs during the asynchronous
+operation, it will receive a single argument - a _Throwable_
+* Promises can be used in a synchronous way to prevent methods that use methods returning promises
+having to return a promise as well ([see explanation](#callback-hell)); This will be done by promises having _wait_ method
+* _wait_ method does not throw an exception if the promise is rejected
```

Comment:
Could you explain what `wait` method does?

Reply:
waits when promise is resolved (resolve method called by initiator). Please see guzzle promises repo for examples